File: tools/vpktool.cpp
#include "core/fs.h"
#include "core/vpk.h"

#include "tool.h"

bool g_verbose;
bool g_debug;

// Initializes stuff like the log level
void Initialize()
{
    spdlog::set_level(spdlog::level::warn);
    if (g_verbose)
    {
        spdlog::set_level(spdlog::level::info);
    }
    if (g_debug)
    {
        spdlog::set_level(spdlog::level::trace);
    }
}

class CreateCommand : public Subcommand
{
  public:
    std::string inputDirectory;
    std::string outputVpk;

    void Run()
    {
        Initialize();

        fs::path directory = fs::absolute(inputDirectory);
        if (!outputVpk.length())
        {
            outputVpk = directory.string();
            if (outputVpk.back() == '\\' || outputVpk.back() == '/')
            {
                outputVpk.resize(outputVpk.length() - 1);
            }
            outputVpk = fmt::format("{}.vpk", outputVpk);
        }

        Core::Vpk::Vpk2 vpk(outputVpk, true);

        fmt::print("Creating VPK file {} from directory {}\n", outputVpk,
                   directory.string());
        for (const auto& entry : fs::recursive_directory_iterator(directory))
        {
            if (entry.is_regular_file() || entry.is_symlink())
            {
                std::string innerPath =
                    fs::relative(entry.path(), directory).string();
                vpk.AddFile(innerPath, Core::Filesystem::Read(entry.path().string()));
            }
        }

        vpk.Write();
    }
};

class ExtractCommand : public Subcommand
{
  public:
    std::string inputVpk;
    std::string outputDirectory;

    void Run()
    {
        Initialize();

        // 4 is length of _dir
        if (inputVpk.length() > 4 + Core::Vpk::VPK_EXTENSION_LENGTH &&
            inputVpk.substr(inputVpk.length() - 4 - Core::Vpk::VPK_EXTENSION_LENGTH,
                            4) == "_dir")
        {
            inputVpk.resize(inputVpk.length() - 4 - Core::Vpk::VPK_EXTENSION_LENGTH);
            inputVpk += Core::Vpk::VPK_EXTENSION;
        }
        fs::path vpkPath = fs::absolute(inputVpk);
        if (!outputDirectory.length())
        {
            outputDirectory = vpkPath.string();
            outputDirectory.resize(outputDirectory.length() -
                                   Core::Vpk::VPK_EXTENSION_LENGTH);
        }
        fs::path directoryPath(outputDirectory);

        fmt::print("Extracting VPK file {} to directory {}\n", vpkPath.string(),
                   directoryPath.string());

        Core::Vpk::Vpk2 vpk(vpkPath.string());
        for (const auto& entry : vpk)
        {
            fs::path path(entry.first);
            path = directoryPath / path;
            fs::create_directories(path.parent_path());
            if (g_verbose)
            {
                fmt::print("Writing file {}\n", path.string());
            }
            Core::Filesystem::Write(path.string(), vpk.Read(entry.first));
        }
    }
};

class ListCommand : public Subcommand
{
  public:
    std::string inputVpk;
    bool checkHashes;

    void Run()
    {
        Initialize();

        // 4 is length of _dir
        if (inputVpk.length() > 4 + Core::Vpk::VPK_EXTENSION_LENGTH &&
            inputVpk.substr(inputVpk.length() - 4 - Core::Vpk::VPK_EXTENSION_LENGTH,
                            4) == "_dir")
        {
            inputVpk.resize(inputVpk.length() - 4 - Core::Vpk::VPK_EXTENSION_LENGTH);
            inputVpk += Core::Vpk::VPK_EXTENSION;
        }
        fs::path vpkPath = fs::absolute(inputVpk);

        Core::Vpk::Vpk2 vpk(vpkPath.string());
        const auto& header = vpk.GetHeader();
        fmt::print("Header:\n");
        fmt::print("\tSignature: {:08X} (should be {:08X})\n", header.signature,
                   Core::Vpk::VPK2_SIGNATURE);
        fmt::print("\tVersion: {}\n", header.version);
        fmt::print("\tTree size: {}\n", header.treeSize);
        fmt::print("\tSize of file data in directory: {}\n",
                   header.fileDataSize);
        fmt::print("\tMD5 section size: {}\n", header.externalMd5Size);
        fmt::print("\tInternal MD5 data size: {}\n", header.md5Size);
        fmt::print("\tSignature data size: {}\n", header.signature);
        fmt::print("\tFile count: {}\n", vpk.GetFileCount());
        for (const auto& pair : vpk)
        {
            const auto& name = pair.first;
            const auto& entry = pair.second;

            fmt::print("Entry {}\n", name);
            if (checkHashes)
            {
                uint32_t crc = vpk.ComputeCrc32(name);
                fmt::print("\tStored CRC: 0x{:08X} (computed 0x{:08X}{})\n",
                           entry.crc, crc,
                           entry.crc != crc ? ", data could be incorrect"
                                            : ", data is correct");
            }
            else
            {
                fmt::print("\tCRC: 0x{:08X}\n", entry.crc);
            }
            fmt::print("\tPreload size: {}\n", entry.preloadSize);
            fmt::print("\tArchive index: {}\n", entry.archiveIndex);
            fmt::print("\tOffset: 0x{:08X}\n", entry.offset);
            fmt::print("\tLength: {}\n", entry.length);
        }
    }
};

int32_t ToolMain()
{
    CLI::App app(fmt::format("{} Valve Pack File (VPK) Tool", Core::GAME_NAME));
    app.set_version_flag("--version",
                         fmt::format("v{}.{}.{}", Core::GAME_MAJOR_VERSION,
                                     Core::GAME_MINOR_VERSION, Core::GAME_PATCH_VERSION));
    app.set_help_all_flag("--help-all", "Extra help");
    app.require_subcommand();

    app.add_flag<bool>("-v,--verbose", g_verbose,
                       "Output information for each file");
    app.add_flag<bool>("-d,--debug", g_debug, "Output debugging information");

    auto create =
        app.add_subcommand("create", "Create a VPK file from a directory");
    CreateCommand createData{};
    create
        ->add_option("-i,--directory", createData.inputDirectory,
                     "The directory to make a VPK file from")
        ->type_name("FILE")
        ->required();
    create
        ->add_option("-o,--vpk", createData.outputVpk,
                     "The path to save the VPK file to (defaults to the name "
                     "of the directory)")
        ->type_name("DIR");
    createData.Register(create);

    auto extract =
        app.add_subcommand("extract", "Extract a VPK file to a directory");
    ExtractCommand extractData{};
    extract
        ->add_option("-i,--vpk", extractData.inputVpk,
                     "The VPK file to extract")
        ->type_name("FILE")
        ->required();
    extract
        ->add_option("-o,--directory", extractData.outputDirectory,
                     "The directory to extract the VPK file into (will be "
                     "created, defaults to the name of the VPK)")
        ->type_name("DIR");
    extractData.Register(extract);

    auto list = app.add_subcommand(
        "list", "List the information and contents of a VPK file");
    ListCommand listData{};
    list->add_option("-i,--vpk", listData.inputVpk, "The VPK file to read")
        ->type_name("FILE")
        ->required();
    list->add_flag("-c,--check", listData.checkHashes,
                   "Check the hashes of files where possible");
    listData.Register(list);

    Core::Filesystem::Initialize();

    CLI11_PARSE(app);

    return 0;
}

File: tools/src/bin/vpktool.rs
use std::{
    io,
    path::{Path, PathBuf},
};

use clap::{Parser, Subcommand};
use common::{
    fs::{self, FileSystem},
    log::{self, error, info},
    vpk,
};

#[derive(Debug, Subcommand)]
enum Commands {
    #[command(about = Some("Create a VPK file from a directory"))]
    Create {
        input_dir: PathBuf,
        output_vpk: Option<PathBuf>,
    },
    #[command(about = Some("Extract a VPK file to directory"))]
    Extract {
        input_vpk: PathBuf,
        output_dir: Option<PathBuf>,
    },
    #[command(about = Some("List the files in a VPK file"))]
    List {
        input_vpk: PathBuf,
        #[arg(short, long)]
        check_hashes: bool,
    },
}

#[derive(Debug, Parser)]
#[command(author, version, about, long_about = Some("Manipulates Valve Pack Files"))]
#[command(propagate_version = true)]
struct Args {
    #[command(subcommand)]
    command: Commands,

    /// Check hashes when listing files
    #[arg(short, long, default_value_t = false)]
    check_hashes: bool,

    /// Be verbose
    #[arg(short, long, default_value_t = false)]
    verbose: bool,

    /// Be extremely verbose
    #[arg(short, long, default_value_t = false)]
    debug: bool,
}

fn create(input_dir: PathBuf, output_vpk: Option<PathBuf>) -> io::Result<()> {
    let mut output_vpk = if output_vpk.is_some() {
        output_vpk.unwrap()
    } else {
        input_dir.clone()
    };
    if output_vpk.ends_with("/") || output_vpk.ends_with("\\") {
        output_vpk.pop();
    }
    output_vpk.set_extension(vpk::VPK_EXTENSION);

    println!(
        "Creating VPK {} from {}",
        input_dir.display(),
        output_vpk.display()
    );

    let mut vpk = vpk::vpk2::Vpk2::new(&output_vpk, true).expect("Failed to create VPK");

    fn visit_dirs<T>(
        dir: &Path,
        cb: &dyn Fn(&fs::DirEntry, &mut Option<T>),
        data: &mut Option<T>,
    ) -> io::Result<()> {
        if dir.is_dir() {
            for entry in std::fs::read_dir(dir)? {
                let entry: fs::DirEntry = entry?.into();
                let path = entry.path();
                if path.is_dir() {
                    visit_dirs(&path, cb, data)?;
                } else {
                    cb(&entry, data);
                }
            }
        }
        Ok(())
    }

    visit_dirs(
        input_dir.as_ref(),
        &|entry: &fs::DirEntry, vpk: &mut Option<&mut vpk::vpk2::Vpk2>| {
            let data = std::fs::read(entry.path())
                .expect(format!("Failed to read file {}", entry.path().display()).as_str());
            let mut path = entry.path();
            path = PathBuf::from(
                path.strip_prefix(input_dir.as_path())
                    .expect("Failed to make path relative"),
            );
            info!("{} -> {}", entry.path().display(), path.display());
            vpk.as_mut()
                .unwrap()
                .write(path.as_path(), data.as_slice())
                .expect(
                    format!("Failed to write file {} into VPK", entry.path().display()).as_str(),
                );
        },
        &mut Some(&mut vpk),
    )
    .expect("Failed to add files to VPK");

    println!("Saving VPK to {}", output_vpk.display());
    vpk.save(output_vpk.as_path()).expect("Failed to save VPK");

    Ok(())
}

fn extract(input_vpk: PathBuf, output_dir: Option<PathBuf>) -> io::Result<()> {
    let base_path = vpk::vpk2::Vpk2::get_base_path(input_vpk.as_path())
        .expect("Failed to get base name of VPK");

    let mut output_dir = if output_dir.is_some() {
        output_dir.unwrap()
    } else {
        base_path.clone()
    };
    output_dir.set_extension("");

    println!(
        "Extracting VPK {} to {}",
        base_path.display(),
        output_dir.display()
    );

    let vpk = vpk::vpk2::Vpk2::new(base_path.as_path(), false).expect("Failed to open VPK");

    for entry in vpk.read_dir(PathBuf::from("").as_path())? {
        match entry {
            Ok(entry) => {
                let mut output_path = output_dir.clone();
                output_path.push(entry.path().file_name().expect("No path"));
                info!("{} -> {}", entry.path().display(), output_path.display());
            }
            Err(err) => {
                error!("Failed to extract file: {err}");
            }
        }
    }

    Ok(())
}

fn main() {
    let args = Args::parse();

    common::setup_logger(
        if args.debug {
            log::LevelFilter::Debug
        } else if args.verbose {
            log::LevelFilter::Info
        } else {
            log::LevelFilter::Warn
        },
        None,
        true,
    )
    .expect("Failed to initialize logging");

    let _ = match args.command {
        Commands::Create {
            input_dir,
            output_vpk,
        } => create(input_dir, output_vpk),
        Commands::Extract {
            input_vpk,
            output_dir,
        } => extract(input_vpk, output_dir),
        Commands::List {
            input_vpk,
            check_hashes,
        } => Ok(()),
    };
}
